internal/wait/pod.go

```
@@ -109,3 +109,49 @@ func (wt Waiter) ForPodListAllRunning(ctx context.Context, pods []*corev1.Pod) (
 	wg.Wait()
 	return failed, updated
 }
+
+func (wt Waiter) ForPodListAllDeleted(ctx context.Context, pods []corev1.Pod) ([]*corev1.Pod, []*corev1.Pod) {
+	var lock sync.Mutex
+	var runningPods []*corev1.Pod
+	var deletedPods []*corev1.Pod
+
+	if len(pods) == 0 {
+		klog.Infof("ForPodListAllDeleted called with an empty list of pods. Nothing to do.")
+		return runningPods, deletedPods
+	}
+
+	klog.Infof("Waiting for %d pod(s) to be deleted.", len(pods))
+
+	var wg sync.WaitGroup
+	for _, pod := range pods {
+		wg.Add(1)
+		go func(currentPod *corev1.Pod) {
+			defer wg.Done()
+
+			podKey := client.ObjectKeyFromObject(currentPod).String()
+			klog.Infof("Goroutine started: waiting for pod %s to be deleted.", podKey)
+			err := wt.ForPodDeleted(ctx, currentPod.Namespace, currentPod.Name)
+
+			lock.Lock()
+			if err != nil {
+				klog.Warningf("Failed to confirm deletion for pod %s: %v", podKey, err)
+				runningPods = append(runningPods, currentPod)
+			} else {
+				klog.Infof("Successfully confirmed deletion for pod %s.", podKey)
+				deletedPods = append(deletedPods, currentPod)
+			}
+			lock.Unlock()
+		}(&pod)
+	}
+
+	wg.Wait()
+
+	klog.Infof("Finished waiting for pod deletions. DeletedPods: %d, RunningPods: %d.", len(deletedPods), len(runningPods))
+	if len(runningPods) > 0 {
+		for _, pod := range runningPods {
+			klog.Warningf("Pod %s/%s was NOT confirmed as deleted.", pod.Namespace, pod.Name)
+		}
+	}
```

I see the point of this, and it is a good addition, except that this can be simpler especially because the nexted ForPodDeleted already reports the undelete pod, and because there is no use of the returned pod lists apart from verifying they fulfill the intention of the function (e.g delete all pods)
instead of waitgroup you can use errgroup.group (https://github.com/openshift-kni/numaresources-operator/blob/main/test/internal/deploy/deploy.go#L150C9-L150C23) and take advantage of the returned error from ForPodDeleted. then at the end of the function return wg.Wait() and outside assert the error expecting it to be nil, fail otherwise. Please correct me if I am missing the point of this.

---

I see what you mean, yes it would be better to return an error. I'll address that, thanks for noting this.